Invite management

zuuring opened the pull request:
### Notes
This adds a new feature to valkyrie for  managing invites into discord both for create and deleting invite.

You can use the `/create-invite` command which will generate an invite for the channel that the command is being sent from. With the default settings of 7 days expiry and 10 uses.

### Defense implementation
This now includes an interaction `discordClient.on("channelCreate")` to check if a channel has been made with `ext-*-audit` format and will generate an invite link into that channel based off this. This will also create a matching role with channel access called `Defense: role`

### Screenshots
<img width="625" alt="Screenshot 2024-03-12 at 15 37 06" src="https://github.com/thesis/valkyrie/assets/7145746/5c25fab2-9a45-432e-bc1b-35f6e1738245">

### Role assignments
Since this is only the component to create/delete invites, we would need to still tie this together with the `role-management.ts` implementation. That being that when a new member joins the discord server, we monitor the `discordClient.on("GuildMemberAdd")` event and check the inviteID used, then assign role tied to that. Whether that should always assume the channel to which an invite is generated a role should be assigned to that once the invite is used (?).

### To-do

- [x] Setup on test discord
- [x] Automated case example, invites on specific channel creation
- [x] Add global command `/create-invite`
- [x] Linking to role assignment (?)

zuuring commented:
Just added the `ext-*-audit` channel invite functionality. Generates an invite to the specific channel based of a channel being made in the defense category with `ext-*-audit` format.

To-do: Link with role-manager to whirl up a role with the same channel name, granting access to that channel only.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,134 @@
+import { Robot } from "hubot"
+import { Client, TextChannel, VoiceChannel, StageChannel } from "discord.js"
+
+async function manageInvite(
+  discordClient: Client,
+  guildId: string,
+  channelId: string,
+  action: "create" | "delete",
```
Let's not get ahead of our skates here—if we're not supporting `delete` yet (a correct decision), let's just call this `createInvite` and skip the action stuff.

Most likely the right way to approach if we do add deleting will be to add a `deleteInvite` and extract the common lookup stuff to a common helper.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,134 @@
+  const { application } = discordClient
+
+  if (application) {
+    discordClient.on("messageCreate", async (message) => {
+      if (message.author.bot || !message.content.startsWith("!create-invite"))
+        return
+
+      if (!message.guild) return
+
+      const channelId = message.channel.id
+      const guildId = message.guild.id
+      const maxAge = 86400 // 1 day
+      const maxUses = 10
+
+      try {
+        const inviteUrl = await manageInvite(
+          discordClient,
+          guildId,
+          channelId,
+          "create",
+          maxAge,
+          maxUses,
+        )
+
+        if (inviteUrl) {
+          robot.logger.info("New invite URL:", inviteUrl)
+          message.channel.send(
+            `Here is your invite link: ${inviteUrl}\n` +
+              `This invite expires in ${
+                maxAge / 3600
+              } hours and has a maximum of ${maxUses} uses.`,
+          )
+        } else {
+          message.channel.send("Unable to create an invite link.")
+        }
+      } catch (error) {
+        message.channel.send(`An error occurred: ${error}`)
+      }
+    })
```
Let's move this out into a proper Discord command.

Note that even if we wanted to do a plain-text message, we should use the Hubot `robot.hear` functionality, as with e.g. https://github.com/thesis/valkyrie/blob/main/discord-scripts/help-me-archive.ts#L21 .

zuuring replied:
@Shadowfiend I've got this setup now in the latest commit with new command `/create-invite` with the same functionality. I also added two parameters to the command `max-age` and `max-uses` that optionally let you specify the maxAge (in hours) or maxUses (number of uses) to the invite. By default it will be set to 24 hours and 10 invites uses. Does that seem like a good feature or not needed at the moment?

Shadowfiend replied:
I would drop it because it'll make things more complicated for folks.

Shadowfiend replied:
That said—I would use a default time range of 7 days and a default max uses of 10 as you've already done.

If we then see there are issues with this, we can revisit.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,134 @@
+        const regex = /^ext-.*-audit$/
+        if (regex.test(channel.name)) {
+          try {
+            const guildId = channel.guild.id
+            const channelId = channel.id
```
Note that we are going out of our way here to pull ids out that we then pass to `manageInvite` and then `manageInvite` resolves them back to first-class objects—let's skip that back and forth and just pass the channel directly to `manageInvite`, which can then resolve the `guild` as a property.

zuuring commented:
Added the permission / role creation in the last commit, so now when `ext-*-audit` channel is created, we create a role with the exact same name and grant it permission to that channel.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+async function createInvite(
+  discordClient: Client,
+  channelId: string,
+  maxAge: number | string = 86400,
```
Let's add relevant constants to globals.ts (I did this already @ https://github.com/thesis/valkyrie/pull/271/files#diff-75c9d9a77a387c5f245414547310fba482857ff1d90b45e0a43381a3ecd25e32 but that PR is a mess so we can't just cherry-pick a commit), and then define this in terms of those constants (e.g. `1 * WEEK`).

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+  maxAge: number | string = 86400,
+  maxUses: number | string = 0,
+): Promise<string | null> {
+  if (!discordClient) {
+    throw new Error("Discord client is not initialized.")
+  }
```
TypeScript guarantees this won't happen, since `discordClient` is of type `Client`, not `Client | undefined`, `Client | null`, or some combo thereof.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+    throw new Error("Discord client is not initialized.")
+  }
+
+  let channelOrigin: TextChannel | VoiceChannel | StageChannel | null = null
+  try {
+    const fetchedChannel = await discordClient.channels.fetch(channelId)
+    if (
+      fetchedChannel instanceof TextChannel ||
+      fetchedChannel instanceof VoiceChannel ||
+      fetchedChannel instanceof StageChannel
+    ) {
+      channelOrigin = fetchedChannel
+    }
+  } catch (error) {
+    throw new Error("Channel not found or access denied.")
+  }
+
+  if (!channelOrigin) {
+    return null
+  }
```
TypeScript's control flow analysis should be able to recognize that this version will work the same way:

```suggestion
  const channelOrigin = await discordClient.channels.fetch(channelId)
  
  if (
    channelOrigin === null ||
    (!(channelOrigin instanceof TextChannel) &&
    !(channelOrigin instanceof VoiceChannel) &&
    !(channelOrigin instanceof StageChannel))
  ) {
   throw new Error("Channel not found or access denied.")
  }
```

This is because TypeScript will see that the above control flow will throw if the fetched channel is null or not one of the 3 types, so if control flows past this statement the channel *must* be one of those three types.

We can actually get even fancier here if we want to though… We can filter only for channels that allow invites:

```suggestion
  const channelOrigin = await discordClient.channels.fetch(channelId)
  
  if (
    channelOrigin === null ||
    !("createInvite" in channelOrigin)
  ) {
   throw new Error("Channel not found or access denied.")
  }
```

I might also consider renaming `channelOrigin` for clarity—perhaps to `channelForInvite` or so?

zuuring replied:
Thanks for the feedback on this and it's a good way to stop my old JS habits, the simplicity of the last commit suggestion is 💯 🤩

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+  if (!channelOrigin) {
+    return null
+  }
+  const numericMaxAge = parseInt(maxAge.toString(), 10)
+  const numericMaxUses = parseInt(maxUses.toString(), 10)
+  if (Number.isNaN(numericMaxAge) || Number.isNaN(numericMaxUses)) {
+    throw new Error("maxAge and maxUses must be valid numbers.")
+  }
```
Even if we did allow these from the command, it's generally a good idea to keep the parsing of string inputs as close to the receiving of those inputs as possible, so that this function can stay focused exclusively on invite creation rather than input processing.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+  const invite = await channelOrigin.createInvite({
+    maxAge: numericMaxAge,
+    maxUses: numericMaxUses,
+    unique: true,
```
Interesting. Wonder what happens if this is off lol.

Shadowfiend replied:
Oh I just meant the `unique` property.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+        !interaction.isCommand() ||
+        interaction.commandName !== "create-invite"
+      )
+        return
```
Really should have curly braces around this—surprised that prettier didn't slap it actually.

Shadowfiend replied:
Flagging this again.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+      try {
+        const { channel } = interaction
+        if (channel instanceof TextChannel) {
+          const invite = await channel.createInvite({
+            maxAge,
+            maxUses,
+            unique: true,
+          })
+          await interaction.reply(
+            `Here is your invite link: ${invite.url}\nThis invite expires in ${
+              maxAge / 3600
+            } hours and has a maximum of ${maxUses} uses.`,
+          )
```
Should this go through the common `createInvite` path?

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+          )
+        }
+      } catch (error) {
+        await interaction.reply("An error occurred while creating an invite.")
```
```suggestion
        await interaction.reply("An error occurred while creating the invite.")
```

At this point they've run `/create-invite`, so it's not just “an” invite, it's the specific one they're trying to create.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+    // Generates an invite if the channel name matches ext-*-audit format
+    discordClient.on("channelCreate", async (channel) => {
+      if (channel.parent && channel.parent.name === "defense") {
+        const regex = /^ext-.*-audit$/
```
Let's pull this up to be a file-level constant (maybe `EXTERNAL_AUDIT_CHANNEL_REGEXP` or so; using `regexp` here because JS uses `RegExp` as the class name).

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+            }
+            // create a new role with the exact name of the channel with permissions only to that channel
+            const role = await channel.guild.roles.create({
+              name: channel.name,
```
The role name here should instead be `Defense: <name of client>`. I suggest we actually rework the `EXTERNAL_AUDIT_CHANNEL_REGEXP` to have a capturing group, so `/^ext-(?<client>.*)-audit$/`. That will let us do, here, `"Defense: ${channel.name.replace(EXTERNAL_AUDIT_CHANNEL_REGEXP, "$< client >")}` or so.

Shadowfiend replied:
Ahh yes; should be fixable with a `/^[^e]+ext-`-type prefix, but I'll have a look and imagine it'll be just fine.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+
+            await channel.permissionOverwrites.create(role, {
+              ViewChannel: true,
+              SendMessages: true,
+              ReadMessageHistory: true,
```
I would nuke these, I believe the only permission overwrite we need in our matrix is `ViewChannel` by design.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -0,0 +1,181 @@
+              SendMessages: true,
+              ReadMessageHistory: true,
+            })
+            if (channel instanceof TextChannel) {
```
I wonder if we should only do this for text channels, especially to start...

Shadowfiend reviewed (commented):
Left a bunch of thoughts. This feels very close to deployable now tbh.

zuuring commented:
@Shadowfiend One thing to note on the `globals.ts` file. I fixed this since `SECOND` was set to 1000 instead of 1, making all day, week, month calculations incorrect., so added in `MILLISECOND = 1000` and updated the ref in `github-auth.ts`. Can retract this change if it's worth saving for another PR.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -1,47 +1,24 @@
   channelId: string,
-  maxAge: number | string = 86400,
-  maxUses: number | string = 0,
+  maxAge = WEEK,
```
Small note that I would tweak this to be `1 * WEEK` just for an easier read.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -93,11 +56,8 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
-        3600
-      const maxUses =
-        (interaction.options.get("max-uses", false)?.value as number) ?? 10
+      const maxAge = WEEK // default 7 days
```
Ditto re `1 * WEEK`—though you can also do `7 * DAYS` or whatever. It's a little weird to have the code say 1 week and the comment say 7 days :)

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -109,71 +69,75 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
             `Here is your invite link: ${invite.url}\nThis invite expires in ${
-              maxAge / 3600
-            } hours and has a maximum of ${maxUses} uses.`,
+              maxAge / (3600 * 24)
```
One of the lovely properties of our constants is this is synonymous with `maxAge / DAYS`, i.e. they can be used both as a multiplier and a divisor.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -109,71 +69,75 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+      ) {
+        try {
+          const channelId = channel.id
+          const maxAge = WEEK
+          const maxUses = 10
```
At this point we've repeated our default age and uses 3 times. I think what I'd suggest here is to make `createInvite` return an object (`{ url: string; maxAge: number; maxUses: number }`) and be entirely in charge of the default age and uses (using its parameter defaults). That way we cut most of this down.

I would also consider making `createInvite` take a `TextChannel` instead of a channel id, since all callers have access to the object. That way you can also eliminate the extra channel lookup by id that we do in `createInvite`.

zuuring replied:
I've got this all fixed now so we only have those defaults on maxAge / maxUses in the main func. Now they are set as they should be as parameter defaults, so if needed we can easily add custom values down the road. And instead of passing channel.id, it just sends the channel.object and runs invite from there.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -109,71 +69,75 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+                } days and has a maximum of ${maxUses} uses.`,
             )
           }
+          // Create a new role with the client name extracted and set permissions to that channel
+          const auditChannel = channel.name.split("-")
+          const clientName = auditChannel[0].includes("🔒")
+            ? auditChannel[1]
+            : ""
```
This feels like it would fall over if e.g. we had a multi-word client name. So for example `ext-tally-ho-audit` would probably result in `Defense: tally` instead of `Defense: Tally Ho` (the ideal outcome).

Let me take the `split` approach one step further and suggest:

```suggestion
          // Create a new role with the client name extracted and set permissions to that channel
          const clientName =
            channel.name
              .split("-")
              // Drop the first and last segments.
              .slice(1, -1)
              // Capitalize each segment.
              .map((segment) => segment.substring(0,1).toUpperCase() + segment.substring(1))
              // Treat the dash separators as spaces.
              .join(" ")
```

In this scenario, I would also check to make sure the `clientName` isn't empty, and if it is just log an error with the channel name and skip the role creation.

Shadowfiend commented on lib/globals.ts:
```diff
@@ -1,3 +1,7 @@
 export const HOST = process.env.HUBOT_HOST
-export const SECOND = 1000
+export const MILLISECOND = 1000
+export const SECOND = 1
```
This should be `1 * MILLISECOND`—if we need to do a division later on by `MILLISECOND` to have the right precision for e.g. the Discord API, let's do that there.

We can probably drop a comment on these indicating that they're in millisecond precision.

The reasoning to do that rather than reindex everything against seconds is:
- More precision is good in many cases, until it's time to drop to lower precision.
- `30 * MILLISECOND` (as used below) isn't correct—we're not setting an interval for 30 milliseconds, but rather for 30 seconds. The goal of the constants here is so that the phrase that uses it reflects the intent, and to do that across most use cases successfully, we need to use the higher precision as the basis.

zuuring replied:
Ok that makes sense! So then for the Discord API, since we need exact seconds for DAY, WEEK instances we will have to do `WEEK / MILLISECOND` in-order to get the desired `maxAge: 604800` vs `604800000` by default. Unless you'd prefer we just add that into the `global.ts` scope, something along the lines of `WEEK_IN_SECONDS`.

Shadowfiend reviewed (changes requested):
Left a handful more comments; the only blocker here is the client name handling for the role—let's get that fixed up and then we can ship, the rest can happen in another PR.
